core/schema/src/main/java/org/hawkular/metrics/schema/SchemaService.java

```diff
@@ -66,51 +65,58 @@ public void run(Session session, String keyspace, boolean resetDB, int replicati
         URI script = getScript();
         cassalog.execute(script, tags, vars);
 
-        session.execute("INSERT INTO " + keyspace + ".sys_config (config_id, name, value) VALUES " +
-                "('org.hawkular.metrics', 'version', '" + getNewHawkularMetricsVersion() + "')");
+        if (updateVersion) {
+            try {
+                updateVersion(session, keyspace, 0, 0);
+            } catch (Exception e) {
+                throw new VersionUpdateException("Failed to update version in system configuration", e);
+            }
+        }
     }
 
-    private URI getScript() {
+    public void updateVersion(Session session, String keyspace, long delay, int maxRetries)
+            throws InterruptedException {
         try {
-            return getClass().getResource("/org/hawkular/schema/cassalog.groovy").toURI();
-        } catch (URISyntaxException e) {
-            throw new RuntimeException("Failed to load schema change script", e);
+            doVersionUpdate(session, keyspace, delay, maxRetries);
+        } catch (Exception e) {
+            throw new VersionUpdateException("Failed to update version in system configuration", e);
         }
     }
 
-    @SuppressWarnings("unused")
-    private boolean systemSettingsTableExists(Session session, String keyspace) {
-        Statement statement = new SimpleStatement("SELECT * FROM sysconfig.schema_columnfamilies WHERE " +
-                "keyspace_name = '" + keyspace + "' AND columnfamily_name = 'system_settings'");
-        ResultSet resultSet = session.execute(statement);
-        return !resultSet.isExhausted();
-    }
+    private void doVersionUpdate(Session session, String keyspace, long delay, int maxRetries)
+        throws InterruptedException {
+        String configId = "org.hawkular.metrics";
+        String configName = "version";
+        String version = VersionUtil.getVersion();
+        PreparedStatement updateVersion = null;
+        int retries = 0;
 
-    @SuppressWarnings("unused")
-    private String getCurrentHawkularMetricsVersion(Session session, String keyspace) {
-        Statement statement = new SimpleStatement("SELECT value FROM " + keyspace + ".sys_config WHERE " +
-                "name = 'org.hawkular.metrics.version'");
-        ResultSet resultSet = session.execute(statement);
-        if (resultSet.isExhausted()) {
-            return null;
+        while (true) {
+            try {
+                if (updateVersion == null) {
```

Comment:
Why not do it before the loop? No point checking here each time if it was prepared? Also, is filling preparedStatementCache useful here? If this is only used in the startup.

Reply:
Yeah, there is no need to do it in the loop. This code is executed in one of two ways, either by calling ```SchemaService.run``` or by directly calling ```SchemaService.updateVersion```. In the former case we know that the schema already exists, and in the latter case we assume that it exists.  I will refactor this code.